### PR TITLE
refactor(core): Major refactor to support read-only permissions

### PR DESCRIPTION
**Pull request by ttomsu**
This was much more work than I expected.

Major changes include:

1. Accounts, Applications, and Service Accounts now each have two new objects:
    * `Resource.Permissions` object that contains its Authorization->Roles (see https://github.com/spinnaker/clouddriver/pull/1544 and https://github.com/spinnaker/front50/pull/219)
    * `Set<Authorization>` that contains the specific authorizations for a particular user.

1. Supports backwards compatible `requiredGroupMembership` specification, granting both `READ` and `WRITE` authorizations. When found, it spits out warning messages, prompting the user to start using the `permissions` method instead.

A little more about point 1: We optimize multi-user resolution by only making 1 call to Front50/Clouddriver for the application/accounts config. These `Account` and `Application` objects must be cloned at the time `Set<Authorization>` is resolved and set for each user. Otherwise,  there would be cross-`Authorization` contamination.

tag https://github.com/spinnaker/fiat/issues/56

### CODE REVIEW

File: fiat-core/src/main/java/com/netflix/spinnaker/fiat/model/resources/BaseAccessControlled.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+package com.netflix.spinnaker.fiat.model.resources;
+
+import com.netflix.spinnaker.fiat.model.Authorization;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.List;
+
+@Slf4j
+abstract class BaseAccessControlled<R extends Resource.AccessControlled> implements Resource.AccessControlled<R> {
```
Comment (jtk54) [inline fiat-core/src/main/java/com/netflix/spinnaker/fiat/model/resources/BaseAccessControlled.java]: What's the plan for cutting support for `requiredGroupMembership`? Would you just remove this class and have the classes that used to extend this just implement the `AccessControlled` interface?

File: fiat-core/fiat-core.gradle [hunk 2]
```diff
@@ -7,10 +7,12 @@ dependencies {
   compileOnly spinnaker.dependency("retrofitJackson")
 
   compileOnly spinnaker.dependency("lombok")
+  compileOnly spinnaker.dependency("slf4jApi")
```
Comment (cfieber) [inline fiat-core/fiat-core.gradle]: seems like compile not compileOnly

File: fiat-core/src/main/java/com/netflix/spinnaker/fiat/model/resources/BaseAccessControlled.java (hunk 1 above)
Comment (ttomsu) [inline fiat-core/src/main/java/com/netflix/spinnaker/fiat/model/resources/BaseAccessControlled.java]: Yeah, pretty much. We can make a call at some point in the future as to when "enough time has passed"

File: fiat-core/fiat-core.gradle (hunk 2 above)
Comment (ttomsu) [inline fiat-core/fiat-core.gradle]: Fixed.

**Comment by ttomsu** [comment]
I've refactored (again) to make `Permissions` immutable, and consumable by by Spring's `ConfigurationProperties` and Jackson's `ObjectMapper`. This was not trivial - it introduces a `Permissions.Builder` class, which Spring can shove values into, and those values get sanitized and locked upon `build()`.

This means we can use a single object across Fiat, Front50, and Clouddriver.

I've also removed a bunch of the `copyWithoutAuthorization` stuff. 

A user's restricted accounts/applications are still stored together at resolution time. However, at request time, the applicable authorizations are granted. Visually this looks like:

Resolution time:
Put in repo this info:
Travis has _some_ sort of access to --> App1, AccountA, AccountB
and has roles --> Role1

Clouddriver - get Travis' permission!
App1.getView(Role1) --> READ
AccountA.getView(Role1) --> READ,WRITE
AccountB.getView(Role1) --> WRITE

**Pull request merged**